source/common/upstream/subset_lb.cc

```diff
@@ -477,20 +480,26 @@ void SubsetLoadBalancer::HostSubsetImpl::update(const HostVector& hosts_added,
   HostVectorSharedPtr hosts(new HostVector());
   HostVectorSharedPtr healthy_hosts(new HostVector());
 
+  // It's possible that hosts_added == original_host_set_.hosts(), e.g.: when
+  // calling refreshSubsets() if only metadata change. If so, we can avoid the
+  // predicate() call.
   for (const auto host : original_host_set_.hosts()) {
-    if (predicate(*host)) {
+    bool host_seen = predicate_added.count(host) == 1;
+    if (host_seen || predicate(*host)) {
       hosts->emplace_back(host);
       if (host->healthy()) {
         healthy_hosts->emplace_back(host);
       }
     }
   }
 
+  // Calling predicate() is expensive since it involves metadata lookups; so we
+  // can take a shortcut and just filter healthy hosts in a follow-up call.
```

Comment:
Sorry just for my understanding, what follow-up call? How come we don't have to do this predicate call here? Maybe make the comment a little more detailed?